cpp/src/Slice/Grammar.y

```diff
@@ -527,50 +665,69 @@ optional
 
     if(tag < 0)
     {
-        unit->error("invalid tag `" + scoped->v + "' for optional");
+        unit->error("invalid tag `" + scoped->v + "'");
     }
 
-    OptionalDefTokPtr m = new OptionalDefTok;
-    m->v.optional = tag >= 0;
+    TaggedDefTokPtr m = new TaggedDefTok;
+    m->v.tagged = tag >= 0;
     m->v.tag = tag;
     $$ = m;
 }
-| ICE_OPTIONAL_OP ')'
+| ICE_TAG_START_COMPAT ')'
 {
-    unit->error("missing tag for optional");
-    OptionalDefTokPtr m = new OptionalDefTok; // Dummy
-    m->v.optional = false;
+    unit->warning(Deprecated, string("The `optional' keyword is deprecated, use `tag' instead"));
+    unit->error("missing tag");
+    TaggedDefTokPtr m = new TaggedDefTok; // Dummy
+    m->v.tagged = false;
     m->v.tag = -1;
     $$ = m;
 }
-| ICE_OPTIONAL
+| ICE_TAG_COMPAT
 {
-    unit->error("missing tag for optional");
-    OptionalDefTokPtr m = new OptionalDefTok; // Dummy
-    m->v.optional = false;
+    unit->warning(Deprecated, string("The `optional' keyword is deprecated, use `tag' instead"));
+    unit->error("missing tag");
+    TaggedDefTokPtr m = new TaggedDefTok; // Dummy
+    m->v.tagged = false;
     m->v.tag = -1;
     $$ = m;
 }
 ;
 
 // ----------------------------------------------------------------------
-optional_type_id
+tagged_type_id
 // ----------------------------------------------------------------------
-: optional type_id
+: tag type_id
 {
-    OptionalDefTokPtr m = OptionalDefTokPtr::dynamicCast($1);
+    TaggedDefTokPtr m = TaggedDefTokPtr::dynamicCast($1);
     TypeStringTokPtr ts = TypeStringTokPtr::dynamicCast($2);
+
+//  OptionalPtr opt = OptionalPtr::dynamicCast(ts->v.first);
```

Comment:
Once this commit is merged in, I'll open an issue to un-comment this once we've added optional (nullability) support to all the language mappings.